Automattic/newspack-network
# feat: content distribution class

1208510338230630-as-1208621948575042/f

Introduces the main class for the Network's content distribution with basic functionality for a post to be distributed across the network according to a specified configuration.

It uses Newspack Network implementation on top of data events for that, so it creates 2 new data events:

## `network_post_updated`

Fires once a distributed post updates, with the following payload:

```json
{
	"site_url": "https://node1.com",
	"post_id": 1,
	"config": {
		"enabled": true,
		"site_urls": [ "https://hub.com", "https://node2.com" ],
		"network_post_id": "1234567890abcdef1234567890abcdef"
	},
	"post_data": {
		"title": "Hello World",
		"date": "2021-01-01 00:00:00",
		"slug": "hello-world",
		"post_type": "post",
		"raw_content": "Content",
		"content": "<p>Content</p>",
		"excerpt": "Excerpt"
	}
}
```

`$post_payload['config']` contains the required information for distribution:
 - `enabled`: Whether distribution is enabled
 - `site_urls`: Sites to distribute the post to
 - `network_post_id`: An md5 composed of site URL and post ID to be unique across the network

The network post ID is used to match an existing linked post to perform updates:

https://github.com/Automattic/newspack-network/blob/2d3f9ca6c2dcfa6ecfcbe06db1896e0fa1c6816c/includes/class-content-distribution.php#L278-L304

## `network_linked_post_inserted`

Fires once a linked post is updated, with the following payload:

```json
{
	"origin": {
		"site_url": "https://node1.com",
		"post_id": 1
	},
	"destination": {
		"site_url": "https://node2.com",
		"post_id": 123,
		"is_unlinked": false
	}
}
```

This event is not yet used but should allow the origin site (or hub) to be aware of its linked posts across the network. This can be for UI or system purposes.

### Testing

1. Make sure you have a working Network setup with at least 1 hub and 2 nodes (let's call them hub, node1, node2)
2. On node1, create a new post with various blocks (paragraph, image, quote, button, HPP...) and grab the post ID
3. Open wp shell and configure the post distribution to hub and node2 by running with the following:

```php
$post_id = 0; // Change to your post ID.
$site_urls = [ 'http://hub.local', 'http://node2.local' ]; // Change to match your setup urls.
Newspack_Network\Content_Distribution::set_post_distribution( $post_id, $site_urls );
```

4. The data event is attached to `wp_after_insert_post`, so you can either hit "Save" on the post or trigger it manually via shell with:

```php
Newspack_Network\Content_Distribution::distribute_post( $post_id );
```

5. Once all the events are dealt with, you should see the post with the correct title, date, and content in both hub and node2. Author and taxonomies are not covered in this PR (see 1208510338230630-as-1208621948575089/f and 1208510338230630-as-1208621948575091/f)
6. Let's unlink the post in node2 by running in node2's shell:

```php
$linked_post_id = 0; // Change to node2's linked post ID
Newspack_Network\Content_Distribution::set_post_unlinked( $linked_post_id );
```

7. In node1, make changes to the post and save
8. Once the events are done, confirm the post has been updated in the hub but remained unchanged in node2
9. Make changes to the post in node2 and save
10. Relink the post in node2:

```php
$linked_post_id = 0; // Change to node2's linked post ID
Newspack_Network\Content_Distribution::set_post_unlinked( $linked_post_id, false );
```

11. Confirm the post in node2 now reflects the most recent content from the distributed post
